sale_general_discount/models/sale.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2018 Tecnativa - Sergio Teruel
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+from odoo.addons import decimal_precision as dp
+from lxml import etree
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    general_discount = fields.Float(
+        digits=dp.get_precision('Discount'),
+        string='Discount (%)',
+    )
+
+    @api.multi
+    @api.onchange('partner_id')
+    def onchange_partner_id(self):
+        for order in self:
+            super(SaleOrder, self).onchange_partner_id()
+            order.general_discount = order.partner_id.sale_discount
+
+    @api.multi
+    @api.onchange('general_discount')
+    def onchange_general_discount(self):
+        with self.env.norecompute():
+            for order in self:
+                order.mapped('order_line').update({
+                    'discount': order.general_discount,
+                })
+        self.recompute()
+
+    @api.model
+    def fields_view_get(self, view_id=None, view_type='form', toolbar=False,
+                        submenu=False):
+        res = super(SaleOrder, self).fields_view_get(
+            view_id, view_type, toolbar, submenu)
+        if view_type == 'form':
+            order_xml = etree.XML(res['arch'])
+            order_line_path = "//field[@name='order_line']"
+            order_line_fields = order_xml.xpath(order_line_path)
+            if order_line_fields:
+                order_line_field = order_line_fields[0]
+                if 'context' not in order_line_field.attrib:
+                    order_line_field.set(
+                        "context", "{'default_discount': general_discount}")
+                else:
+                    ctx_orig = order_line_field.attrib['context']
+                    if ctx_orig[-2] == ",":
+                        char_replace = ',}'
+                    else:
+                        char_replace = '}'
```

Comment:
If the answer to previous question is _no_, let me recommend you a shorter and less error-prone, approach:

```python
context = order_line_field.attrib.get("context", "{}")
context = context.replace("{", "{'default_discount': general_discount,", 1)
order_line_field.attrib['context'] = context
```

This should work, no matter the whitespaces and commas you can find, and it's easier to read IMHO.